src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs

```diff
@@ -108,26 +104,20 @@ private void DisposeSyncHelper()
 
             _sendBuffer.Dispose();
             _recvBuffer.Dispose();
-
-            // Dispose() might be called concurrently with GoAway(), we need to make sure to not Dispose/Cancel the CTS concurrently.
-            Interlocked.Exchange(ref _goawayCancellationSource, null)?.Dispose();
         }
 
         public void GoAway()
         {
-            // Dispose() might be called concurrently with GoAway(), we need to make sure to not Dispose/Cancel the CTS concurrently.
-            using CancellationTokenSource? cts = Interlocked.Exchange(ref _goawayCancellationSource, null);
-            cts?.Cancel();
+            _requestBodyCancellationSource.Cancel();
         }
 
         public async Task<HttpResponseMessage> SendAsync(CancellationToken cancellationToken)
         {
             // If true, dispose the stream upon return. Will be set to false if we're duplex or returning content.
             bool disposeSelf = true;
 
-            // Link the input token with _resetCancellationTokenSource, so cancellation will trigger on GoAway() or Abort().
-            using CancellationTokenSource requestCancellationSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _goawayCancellationToken);
-            CancellationTokenRegistration sendContentCancellationRegistration = default;
+            // Link the input token with _requestBodyCancellationSource, so cancellation will trigger on GoAway() or Abort().
+            CancellationTokenRegistration linkedTokenRegistration = cancellationToken.UnsafeRegister(cts => ((CancellationTokenSource)cts!).Cancel(), _requestBodyCancellationSource);
```

Comment:
If I get this right, the cancellation of `cancellationToken` passed into `HttpClient.SendAsync` will trigger cancellation of `_requestBodyCancellationSource` regardless of being duplex or not.

Whereas in H/2 I see that we ignore the user token for duplex contents:
https://github.com/dotnet/runtime/blob/35f66ed7cb42a3fbcc2c56604c31f8a64dbe119b/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L1911-L1914

If the goal of this PR to have parity with H/2, I think we should take duplex into account here as well. But I might be missing something, so please correct me if I'm wrong.

Reply:
You're right, I missed this.

Reply:
> If I get this right, the cancellation of cancellationToken passed into HttpClient.SendAsync will trigger cancellation of _requestBodyCancellationSource regardless of being duplex or not.

Yes, but only before SendAsync returns (we dispose the registration).
Http2 effectively does the same thing - if the cancellation token is signaled within SendAsync, it will cancel the ReadResponseHeaders, which will in turn cancel the content.

I think this may be fine as-is, the logic is just simpler here since we're not split between 2 classes (connection & stream) and the cancellation control is therefore more explicit.
With H2 we can't pass the CT directly into SendRequestContent as the content sending may outlive the SendAsync CT. With H3, we're already using a separate CTS, so we can control the linked registration lifetime.

Reply:
Yeah, H2 does it, though in a very roundabout way...
BTW we've already fixed the problem when a token was (wrongly) conditional before here https://github.com/dotnet/runtime/issues/57619

Reply:
>  it will cancel the ReadResponseHeaders, which will in turn cancel the content.

Baah, I missed that.

Reply:
Seems like we want to keep the registration unconditional. so I reverted the change I proactively did about that :D